File: engineering/concurrency/04-producer-consumer/skeleton/go/main.go
package main

import (
	"fmt"
	"sync"
)

// BoundedBuffer 有界缓冲区。
// TODO: 添加 buffer 切片、头尾索引、计数、互斥锁和条件变量。
type BoundedBuffer struct {
	mu sync.Mutex
	// TODO
}

func NewBoundedBuffer(capacity int) *BoundedBuffer {
	// TODO: 初始化
	return &BoundedBuffer{}
}

// Produce 向缓冲区放入一个元素。若缓冲区满，阻塞等待。
func (b *BoundedBuffer) Produce(item int) {
	b.mu.Lock()
	defer b.mu.Unlock()
	// TODO: 等待 notFull，放入元素，通知 notEmpty
}

// Consume 从缓冲区取出一个元素。若缓冲区空，阻塞等待。
func (b *BoundedBuffer) Consume() int {
	b.mu.Lock()
	defer b.mu.Unlock()
	// TODO: 等待 notEmpty，取出元素，通知 notFull
	return 0
}

func main() {
	const capacity = 10
	const numProducers = 5
	const numConsumers = 3
	const itemsPerProducer = 100

	buffer := NewBoundedBuffer(capacity)
	var wg sync.WaitGroup
	var sum int64
	var mu sync.Mutex

	// 生产者
	for i := 0; i < numProducers; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			for j := 0; j < itemsPerProducer; j++ {
				item := id*itemsPerProducer + j
				buffer.Produce(item)
			}
		}(i)
	}

	// 消费者
	for i := 0; i < numConsumers; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			for j := 0; j < (numProducers*itemsPerProducer)/numConsumers; j++ {
				item := buffer.Consume()
				mu.Lock()
				sum += int64(item)
				mu.Unlock()
			}
		}(i)
	}

	wg.Wait()

	expectedSum := int64(numProducers*itemsPerProducer-1) * int64(numProducers*itemsPerProducer) / 2
	if sum == expectedSum {
		fmt.Printf("PASS: sum = %d (expected %d)\n", sum, expectedSum)
	} else {
		fmt.Printf("FAIL: sum = %d (expected %d)\n", sum, expectedSum)
	}
}

File: engineering/concurrency/04-producer-consumer/skeleton/python/main.py
# 并发模型：threading
import threading
from collections import deque

class BoundedBuffer:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.buffer = deque()
        # TODO: 添加互斥锁和两个条件变量（not_full, not_empty）

    def produce(self, item):
        # TODO: 获取锁；若缓冲区满，等待 not_full；放入元素；通知 not_empty；释放锁
        pass

    def consume(self):
        # TODO: 获取锁；若缓冲区空，等待 not_empty；取出元素；通知 not_full；释放锁
        return 0

def producer(buffer, pid, count):
    for i in range(count):
        buffer.produce(pid * count + i)

def consumer(buffer, count, result, mu):
    s = 0
    for _ in range(count):
        s += buffer.consume()
    with mu:
        result[0] += s

def main():
    capacity = 10
    num_producers = 5
    num_consumers = 3
    items_per_producer = 100

    buffer = BoundedBuffer(capacity)
    threads = []
    result = [0]
    mu = threading.Lock()

    for i in range(num_producers):
        t = threading.Thread(target=producer, args=(buffer, i, items_per_producer))
        threads.append(t)
        t.start()

    items_per_consumer = (num_producers * items_per_producer) // num_consumers
    for i in range(num_consumers):
        t = threading.Thread(target=consumer, args=(buffer, items_per_consumer, result, mu))
        threads.append(t)
        t.start()

    for t in threads:
        t.join()

    total_items = num_producers * items_per_producer
    expected = (total_items - 1) * total_items // 2
    if result[0] == expected:
        print(f"PASS: sum = {result[0]} (expected {expected})")
    else:
        print(f"FAIL: sum = {result[0]} (expected {expected})")

if __name__ == "__main__":
    main()
